aana/processors/speaker.py

```diff
@@ -0,0 +1,633 @@
+from collections import defaultdict
+from typing import TypedDict
+
+from aana.core.models.asr import (
+    AsrSegment,
+    AsrTranscription,
+    AsrTranscriptionInfo,
+    AsrWord,
+)
+from aana.core.models.speaker import SpeakerDiarizationSegment
+from aana.core.models.time import TimeInterval
+
+# Utility functions for speaker-related processing in audio
+
+
+# Redefine SpeakerDiarizationOutput and WhisperOutput to prevent circular imports
+class SpeakerDiarizationOutput(TypedDict):
+    """The output of the Speaker Diarization model.
+
+    Attributes:
+        segments (list[SpeakerDiarizationSegment]): The Speaker Diarization segments.
+    """
+
+    segments: list[SpeakerDiarizationSegment]
+
+
+class WhisperOutput(TypedDict):
+    """The output of the whisper model.
+
+    Attributes:
+        segments (list[AsrSegment]): The ASR segments.
+        transcription_info (AsrTranscriptionInfo): The ASR transcription info.
+        transcription (AsrTranscription): The ASR transcription.
+    """
+
+    segments: list[AsrSegment]
+    transcription_info: AsrTranscriptionInfo
+    transcription: AsrTranscription
+
+
+# Define sentence ending punctuations to split segments at sentence endings:
+sentence_ending_punctuations = ".?!"
+
+
+def assign_word_speakers(
+    diarized_output: SpeakerDiarizationOutput,
+    transcription: WhisperOutput,
+    fill_nearest: bool = False,
+) -> WhisperOutput:
+    """Assigns speaker labels to each segment and word in the transcription based on diarized output.
+
+    Args:
+        diarized_output (SpeakerDiarizationOutput): Contains speaker diarization segments.
+        transcription (WhisperOutput): Transcription data with segments, text, and language_info.
+        fill_nearest (bool): If True, assigns the closest speaker even if there's no positive overlap. Default is False.
+
+    Returns:
+        transcription (WhisperOutput): Transcription updated in-place with the assigned speaker labels.
+    """
+    for segment in transcription["segments"]:
+        # Assign speaker to segment
+        segment.speaker = get_speaker_for_interval(
+            diarized_output["segments"],
+            segment.time_interval.start,
+            segment.time_interval.end,
+            fill_nearest,
+        )
+
+        # Assign speakers to words within the segment
+        if segment.words:
```

Comment:
I saw `whisper_params.word_timestamps = True` in the notebook. As I understand none of this will work without word level predictions. Then we should add check somewhere to make sure that word-level predictions are included. Maybe we can check segment text and if it's not empty but words are empty list then we throw exception.